src/vs/workbench/contrib/mcp/browser/mcpResourceQuickAccess.ts

```diff
@@ -255,19 +329,28 @@ export class McpResourcePickHelper {
 		}).finally(() => store.dispose());
 	}
 
-	public getPicks(onChange: (value: Map<IMcpServer, (IMcpResourceTemplate | IMcpResource)[]>) => void, token?: CancellationToken) {
+	public getPicks(token?: CancellationToken): IObservable<Map<IMcpServer, (IMcpResourceTemplate | IMcpResource)[]>> {
 		const cts = new CancellationTokenSource(token);
-		const store = new DisposableStore();
-		store.add(toDisposable(() => cts.dispose(true)));
-
+		this._register(toDisposable(() => cts.dispose(true)));
 		// We try to show everything in-sequence to avoid flickering (#250411) as long as
 		// it loads within 5 seconds. Otherwise we just show things as the load in parallel.
 		let showInSequence = true;
-		store.add(disposableTimeout(() => {
+		this._register(disposableTimeout(() => {
 			showInSequence = false;
 			publish();
 		}, 5_000));
 
+		this._register(autorun(reader => {
+			const directoryResource = this._inDirectory.read(reader);
+			if (directoryResource) {
+				// When navigating into a directory, filter to show only that server's resources
+				const filtered = new Map<IMcpServer, (IMcpResourceTemplate | IMcpResource)[]>();
+				filtered.set(directoryResource.server, directoryResource.resources);
+				this._resources.set(filtered, undefined);
```

Comment:
This usage of _resources is racey because you could navigate to a subdirectory before all resources have been populated.

Setting another observable from inside an autorun is a code smell, in general you should `derive` the necessary output instead